android_app/app/src/main/java/com/health/openscale/core/bluetooth/BluetoothStandardWeightProfile.java

```diff
@@ -213,91 +387,465 @@ private void handleBodyCompositionMeasurement(byte[] value) {
+        }
 
         // Read height if present
         if (heightPresent) {
             float heightValue = parser.getIntValue(BluetoothBytesParser.FORMAT_UINT16);
+            Timber.d(prefix + "heightValue: " + heightValue);
+        }
+
+        if (multiPacketMeasurement) {
+            Timber.e(prefix + "multiPacketMeasurement not supported!");
         }
 
         Timber.d(String.format("Got body composition: %s", byteInHex(value)));
-        addScaleMeasurement(scaleMeasurement);
+        return scaleMeasurement;
     }
 
-    private void registerUser(int consentCode) {
+    protected void handleBodyCompositionMeasurement(byte[] value) {
+        mergeWithPreviousScaleMeasurement(bodyCompositionMeasurementToScaleMeasurement(value));
+    }
+
+    /**
+     * Bluetooth scales usually implement both "Weight Scale Feature" and "Body Composition Feature".
+     * It seems that scale first transmits weight measurement (with user index and timestamp) and
+     * later transmits body composition measurement (without user index and timestamp).
+     * If previous measurement contains user index and new measurements does not then merge them and
+     * store as one.
+     * disconnect() function must store previousMeasurement to openScale db (if present).
+     *
+     * @param newMeasurement the scale data that should be merged with previous measurement or
+     *                       stored as previous measurement.
+     */
+    protected void mergeWithPreviousScaleMeasurement(ScaleMeasurement newMeasurement) {
+        if (previousMeasurement == null) {
+            if (newMeasurement.getUserId() == -1) {
+                addScaleMeasurement(newMeasurement);
+            }
+            else {
+                previousMeasurement = newMeasurement;
+            }
+        }
+        else {
+            if ((newMeasurement.getUserId() == -1) && (previousMeasurement.getUserId() != -1)) {
+                previousMeasurement.merge(newMeasurement);
+                addScaleMeasurement(previousMeasurement);
+                previousMeasurement = null;
+            }
+            else {
+                addScaleMeasurement(previousMeasurement);
+                if (newMeasurement.getUserId() == -1) {
+                    addScaleMeasurement(newMeasurement);
+                    previousMeasurement = null;
+                }
+                else {
+                    previousMeasurement = newMeasurement;
+                }
+            }
+        }
+    }
+
+    @Override
+    public void disconnect() {
+        if (previousMeasurement != null) {
+            addScaleMeasurement(previousMeasurement);
+            previousMeasurement = null;
+        }
+        super.disconnect();
+    }
+
+    protected void setNotifyVendorSpecificUserList() {
+        Timber.d("Set notify vendor user list not implemented!");
+    }
+
+    protected void requestVendorSpecificUserList() {
+        Timber.d("Request vendor user list not implemented!");
+    }
```

Comment:
and I think, we need to resume it here for all models without a user list:
```suggestion
    resumeMachineState();
    }
```

Reply:
So far all scales (that I have found) that implement Bluetooth Standard Weight Profile - they all have vendor-specific user lists.
I think maybe we should make class BluetoothStandardWeightProfile abstract and force extender classes to implement requestVendorSpecificUserList.
And resuming state inside requestVendorSpecificUserList() won't help if you stop machine before after you call requestVendorSpecificUserList();

Reply:
Oh, yes of course not! My Bad. So maybe it would be better to first stop the machine and then call the function(s) that let the machine resume ...
But your idea to make BluetoothStandardWeightProfile abstract sounds good to me, too.

Reply:
I decided to make BluetoothStandardWeightProfile abstract. I prefer not to write code for hypothetical device  that I have no ability to test.